File: frontend/projects/valtimo/shared/src/lib/validators/bsn-validator.ts
/*
 * Copyright 2015-2025 Ritense BV, the Netherlands.
 *
 * Licensed under EUPL, Version 1.2 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * https://joinup.ec.europa.eu/collection/eupl/eupl-text-eupl-12
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" basis,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { AbstractControl, ValidationErrors, ValidatorFn } from '@angular/forms';

export function bsnValidator(): ValidatorFn {
  return (control: AbstractControl): ValidationErrors | null => {
    const value = control.value?.toString().trim();

    if (!value) return null;

    // Can only contain digits
    if (!/^\d+$/.test(value)) {
      return { invalidFormat: true };
    }

    // Verify the length
    if (value.length < 8 || value.length > 9) {
      return { invalidLength: true };
    }

    // Cannot be full of zeros or ones
    if (/^0+$/.test(value) || /^9+$/.test(value)) {
      return { invalidSequence: true };
    }

    // BSN check digit validation (11-test)
    const digits = value.split('').map(d => parseInt(d, 10));
    let sum = 0;
    const length = digits.length;

    for (let i = 0; i < length; i++) {
      const weight = (i === length - 1) ? -1 : (length - i);
      sum += digits[i] * weight;
    }

    if (sum % 11 !== 0) {
      return { invalidBsn: true };
    }

    return null;
  };
}

File: backend/contract/src/main/kotlin/com/ritense/valtimo/contract/validation/BsnValidator.kt
/*
 * Copyright 2015-2025 Ritense BV, the Netherlands.
 *
 * Licensed under EUPL, Version 1.2 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * https://joinup.ec.europa.eu/collection/eupl/eupl-text-eupl-12
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" basis,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.ritense.valtimo.contract.validation

object BsnValidator {

    private val elfProefWeights = intArrayOf(9, 8, 7, 6, 5, 4, 3, 2, -1)
    private val bsnRegex = Regex("^\\d{8,9}$")

    fun isValid(bsn: String?): Boolean {
        if (bsn.isNullOrBlank()) return false
        if (!bsnRegex.matches(bsn)) return false
        if (bsn.all { it == '0' }) return false

        val normalizedBsn = bsn.padStart(9, '0')
        return passesElfProef(normalizedBsn)
    }

    private fun passesElfProef(value: String): Boolean {
        var result = 0
        for (i in elfProefWeights.indices) {
            val digit = value[i] - '0'
            result += digit * elfProefWeights[i]
        }
        return result % 11 == 0
    }
}
